### PR TITLE
SOPHI-82: First version of device detector for scala

### PR DESCRIPTION
**Pull request by wisdomjiang**
This is my first version for this device detector.  There are a lot more works to do for this project:

1. scalalize everything so that it is a good functional programming style as well as object oriented
2. Unit test everything to make sure code is running well and with good performances.
3. Document everything so that it is ready to be released

We will create a new ticket to finish all of this.

### CODE REVIEW

File: src/main/scala/com/theglobeandmail/devicedetector/parser/ParserAbstract.scala [hunk 1]
```diff
@@ -0,0 +1,207 @@
+package com.theglobeandmail.devicedetector.parser
+
+import com.theglobeandmail.devicedetector.cache.Cache
+import com.theglobeandmail.devicedetector.DeviceDetector
+import com.theglobeandmail.devicedetector.results.ParseResult
+import com.theglobeandmail.devicedetector.yaml.YamlParser
+import com.typesafe.config.ConfigFactory
+
+
+/** regexList: Holds an array with regexes to parse, if already loaded */
```
Comment (szareiangm) [inline src/main/scala/com/theglobeandmail/devicedetector/parser/ParserAbstract.scala]: Please use scaladoc. in your intellij press '/**' and press enter for autocompete

File: src/main/scala/com/theglobeandmail/devicedetector/parser/ParserAbstract.scala [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    parserName
+  }
+
+  /** Returns the result of the parsed yml file defined in $fixtureFile
+    *
+    * @return array
+    */
+  protected def getRegexes: Option[List[A]] = {
+    if (regexList.isEmpty || regexList.get.isEmpty) {
+      val keyForCache = "DeviceDetctor-" + DeviceDetector.VERSION + "regexes-" + getName()
```
Comment (szareiangm) [inline src/main/scala/com/theglobeandmail/devicedetector/parser/ParserAbstract.scala]: is `"DeviceDetctor-" + DeviceDetector.VERSION + ` necessary?

File: src/main/scala/com/theglobeandmail/devicedetector/parser/client/BrowserParser.scala [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  }
+}
+
+object BrowserParser{
+  /**
+    * Known browsers mapped to their internal short codes
+    *
+    * @var Map
+    */
+  val availableBrowsers = Map(
```
Comment (szareiangm) [inline src/main/scala/com/theglobeandmail/devicedetector/parser/client/BrowserParser.scala]: you can move all of these enums to `package` files.

File: src/main/scala/com/theglobeandmail/devicedetector/parser/client/ClientParserAbstract.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    *
+    * @return array
+    */
+  def getAvailableClients: List[A] = {
+    getRegexes.getOrElse(List.empty[A])
+      .filter(r => r.name != "$1")
+      .groupBy(_.name)
+      .map(_._2.head)
+      .toList
+      .sorted
```
Comment (szareiangm) [inline src/main/scala/com/theglobeandmail/devicedetector/parser/client/ClientParserAbstract.scala]: no sort needed. right?

File: src/main/scala/com/theglobeandmail/devicedetector/regexes/client/browser_engine.yml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+###############
```
Comment (szareiangm) [inline src/main/scala/com/theglobeandmail/devicedetector/regexes/client/browser_engine.yml]: could you please move the static files to `resources`?

Comment (wisdomjiang) [reply src/main/scala/com/theglobeandmail/devicedetector/regexes/client/browser_engine.yml]: OK, done

File: src/main/scala/com/theglobeandmail/devicedetector/parser/client/ClientParserAbstract.scala (hunk 4 above)
Comment (wisdomjiang) [reply src/main/scala/com/theglobeandmail/devicedetector/parser/client/ClientParserAbstract.scala]: Probably not I think.  No sort here does not hurt.  Let me remove this and I will confirm in the future test.

File: src/main/scala/com/theglobeandmail/devicedetector/parser/client/BrowserParser.scala (hunk 3 above)
Comment (wisdomjiang) [reply src/main/scala/com/theglobeandmail/devicedetector/parser/client/BrowserParser.scala]: Thanks for the suggestions.  They are awesome.

File: src/main/scala/com/theglobeandmail/devicedetector/parser/ParserAbstract.scala (hunk 2 above)
Comment (wisdomjiang) [reply src/main/scala/com/theglobeandmail/devicedetector/parser/ParserAbstract.scala]: I don't like this too.  The DeviceDetector.Version is only a version number.  As part of my plan, I would need to remove this cache.  And all the constant parameters in different files will be put into one parameter files (probably in yaml or in an object class)

File: src/main/scala/com/theglobeandmail/devicedetector/parser/ParserAbstract.scala (hunk 1 above)
Comment (wisdomjiang) [reply src/main/scala/com/theglobeandmail/devicedetector/parser/ParserAbstract.scala]: Thanks!  I use scaladoc here and in Devicedector.scala, I will do more documentations according to scaladoc style later as a final step to wrap everything up.

**Pull request merged**